native/dtrexc.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	// Determine the first row of specified block and find out it is 1×1 or 2×2.
+	if ifst > 0 && t[ifst*ldt+ifst-1] != 0 {
+		ifst--
+	}
+	nbf := 1 // Size of the first block.
+	if ifst+1 < n && t[(ifst+1)*ldt+ifst] != 0 {
+		nbf = 2
+	}
+	// Determine the first row of the final block
+	// and find out it is 1×1 or 2×2.
+	if ilst > 0 && t[ilst*ldt+ilst-1] != 0 {
+		ilst--
+	}
+	nbl := 1 // Size of the last block.
+	if ilst+1 < n && t[(ilst+1)*ldt+ilst] != 0 {
+		nbl = 2
+	}
+
+	switch {
+	case ifst == ilst:
+		return ifst, ilst, true
+
+	case ifst < ilst:
+		// Update ilst.
+		if nbf == 2 && nbl == 1 {
+			ilst--
+		}
+		if nbf == 1 && nbl == 2 {
+			ilst++
+		}
+		here := ifst
+		for here < ilst {
+			// Swap block with next one below.
+
+			if nbf == 1 || nbf == 2 {
+				// Current block either 1×1 or 2×2.
+				nbnext := 1 // Size of the next block.
+				if here+nbf+1 < n && t[(here+nbf+1)*ldt+here+nbf] != 0 {
+					nbnext = 2
+				}
+				ok = impl.Dlaexc(wantq, n, t, ldt, q, ldq, here, nbf, nbnext, work)
+				if !ok {
+					return ifst, here, false
+				}
+				here += nbnext
+				// Test if 2×2 block breaks into two 1×1 blocks.
+				if nbf == 2 && t[(here+1)*ldt+here] == 0 {
+					nbf = 3
+				}
+				continue
+			}
+
+			// Current block consists of two 1×1 blocks each of
+			// which must be swapped individually.
+			nbnext := 1 // Size of the next block.
+			if here+3 < n && t[(here+3)*ldt+here+2] != 0 {
+				nbnext = 2
+			}
+			ok = impl.Dlaexc(wantq, n, t, ldt, q, ldq, here+1, 1, nbnext, work)
+			if !ok {
+				return ifst, here, false
+			}
+			if nbnext == 1 {
+				// Swap two 1×1 blocks, no problems possible.
+				impl.Dlaexc(wantq, n, t, ldt, q, ldq, here, 1, nbnext, work)
+				here++
+				continue
+			}
+			// Recompute nbnext in case 2×2 split.
+			if t[(here+2)*ldt+here+1] == 0 {
+				nbnext = 1
+			}
+			if nbnext == 2 {
+				// 2×2 block did not split.
+				ok = impl.Dlaexc(wantq, n, t, ldt, q, ldq, here, 1, nbnext, work)
+				if !ok {
+					return ifst, here, false
+				}
+			} else {
+				// 2×2 block did split.
+				impl.Dlaexc(wantq, n, t, ldt, q, ldq, here, 1, 1, work)
+				impl.Dlaexc(wantq, n, t, ldt, q, ldq, here+1, 1, 1, work)
+			}
+			here += 2
+		}
+		return ifst, here, true
+
+	default: // ifst > ilst
+		here := ifst
+		for here > ilst {
+			// Swap block with next one above.
+
+			if nbf == 1 || nbf == 2 {
+				// Current block either 1×1 or 2×2.
+				nbnext := 1
+				if here-2 >= 0 && t[(here-1)*ldt+here-2] != 0 {
+					nbnext = 2
+				}
+				ok = impl.Dlaexc(wantq, n, t, ldt, q, ldq, here-nbnext, nbnext, nbf, work)
+				if !ok {
+					return ifst, here, false
+				}
+				here -= nbnext
+				// Test if 2×2 block breaks into two 1×1 blocks.
+				if nbf == 2 && t[(here+1)*ldt+here] == 0 {
+					nbf = 3
+				}
+				continue
+			}
+
+			// Current block consists of two 1×1 blocks each of
+			// which must be swapped individually.
+			nbnext := 1
+			if here-2 >= 0 && t[(here-1)*ldt+here-2] != 0 {
```

Comment:
Again, I'm undecided. I started using this form after recent @btracey's suggestion. In that case the index was more complex but I started to like this form because it directly links to the index used in `t[...]` (the smaller index must be non-negative). It also helps me in translating from Fortran. For the moment I'm leaving it.